Results_Engine/Query/DisplayElement1DResults.cs

```diff
@@ -0,0 +1,282 @@
+            for (int i = 0; i < indexRanges.Count - 1; i++)
+            {
+                if (indexRanges[i].Item3 == indexRanges[i + 1].Item3)   //If next range has the same colour index as current
+                {
+                    //Replace current range index with one that goes from current to next
+                    indexRanges[i] = new Tuple<double, double, int>(indexRanges[i].Item1, indexRanges[i + 1].Item2, indexRanges[i].Item3);
+                    indexRanges.RemoveAt(i + 1);    //Remove next range as included in current
+                    i--;    //Reduce indexer to chech current at next step as well
+                }
+            }
+            //Finaly, return ranges of startPos, endPos and colour by accecing the colours from the provided markers
+            return indexRanges.Select(x => new Tuple<double, double, Color>(x.Item1, x.Item2, markers[x.Item3].Item2)).ToList();
+        }
+
+        /***************************************************/
+
+        private static Tuple<double, double, int> GetInterpolationRange(double k, double nextStepMarker, double firstVal, double firstPos, double nextPos, int index, ref double startPos)
+        {
+            double endPos = k * (nextStepMarker - firstVal) + firstPos;
+            endPos = Math.Min(endPos, nextPos);
+            var range = new Tuple<double, double, int>(startPos, endPos, index);
+            startPos = endPos;
+            return range;
+        }
+
+        /***************************************************/
+
+        private static int GetColourIndex(double val, List<Tuple<double, Color>> markers)
+        {
+            //Gets the index of the last marker in the list that has a value larger or equal to the value
+            int colIndex = markers.FindLastIndex(x => x.Item1 <= val);
+
+            //If no index is found it means the value is lower than the first item in the marker.
+            //For this case, the lowest index marker is used.
+            if (colIndex == -1)
+                colIndex = 0;
+
+            return colIndex;
+        }
+
+        /***************************************************/
+
+        private static List<List<T>> GroupResults<T>(this IEnumerable<T> results) where T : IResult
+        {
+            if (results.FirstOrDefault() is ICasedResult)
+            {
+                if (results.FirstOrDefault() is ITimeStepResult)
+                    return results.GroupBy(x => new { ((ICasedResult)x).ResultCase, ((ITimeStepResult)x).TimeStep }).Select(x => x.ToList()).ToList();
+                else
+                    return results.GroupBy(x => ((ICasedResult)x).ResultCase).Select(x => x.ToList()).ToList();
+            }
+            else if (results.FirstOrDefault() is ITimeStepResult)
+                return results.GroupBy(x => ((ITimeStepResult)x).TimeStep).Select(x => x.ToList()).ToList();
+            else
+                return new List<List<T>> { results.ToList() };
+
+        }
```

Comment:
I'm thinking we could expose these as public overloaded methods for different groupings. Could be very useful in other instances too.
E.g.
```cs
public List<List<T>> GroupByTimeStep<T>(this IEnumerable<T> results) where T : ITimeStepResult
{
  // do stuff
}

public List<List<T>> GroupByResultCase<T>(this IEnumerable<T> results) where T : ICasedResult
{
  // do stuff
}

public List<List<T>> GroupByResultCaseAndTimeStep<T>(this IEnumerable<T> results) where T : ICasedResult, ITimeStepResult
{
  // do stuff
}
```

Reply:
We actually have the first two methods already :)

For this one, I tried to think of a way to basically group by non-object related identifiers in a generic way, but then run a bit short of coming up with something at the spot.

Would like to have a method like that basically (GroupByNonObejctIdentifiers) or similar. Will have a bit of a think if I can think of something better. Agree it would be useful, and agree this method feels kind of a bit misplaced.

If I cant resolve it quickly, would you be fine with keeping this in this PR if I raise an issue for it to be fixed properly in a further iteration?

Reply:
To summarise in our discussion, this is more about having public methods to get a ILookup of results as you are doing here

Reply:
Done